FlashIAP & NVStore tests: Skip test if overwriting code in flash

davidsaada opened the pull request:
### Description

The FlashIAP test erases the last two sectors in some of its test cases. NVStore test does the same. However, with boards that have small flash components (and large last sectors), this test may overwrite the code section. This fix checks that scenario and skips the test if it is indeed the case. 
Resolves [#7149](https://github.com/ARMmbed/mbed-os/issues/7149) and [#7388](https://github.com/ARMmbed/mbed-os/issues/7388).

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

Commit:
Fix TMPM64B IAR linker file

Remove unnecessary manual inclusion of tmpm64b_fc object file in linker script

Commit:
FlashIAP & NVStore tests: Skip test if overwriting code in flash.

davidsaada commented:
@jeromecoutant Would be glad if you could test this fix on your board, as I don't have it. If you can do it on all supported three toolchains (GCC_ARM, ARM, IAR), it would be best. Note that if it works, you should see a print telling the test was skipped.

0xc0170 commented on TESTS/mbed_drivers/flashiap/main.cpp:
```diff
@@ -28,6 +28,30 @@
+#endif
+
+// This macro returns if flash test overwrites the code in flash.
+#define RETURN_IF_TEST_OVERWRITES_CODE(ADDRESS) \
```
this would be nicer if its just a function, no need for a macro?

davidsaada replied:
I'm not a big macro fan myself. However, in this case, macro seems like the better solution, as it includes a `return` inside. Turning this into a function would make the code a bit more bloated, as it would make all three tests check the return value, and return if it indicates that the code is getting overwritten. No biggie, can change if required.

0xc0170 replied:
👍  Fine as it is

jeromecoutant commented:
Hi
It looks OK with F410 and GCC,

But I couldn't compile for L475VG and ARM...

[DEBUG] Errors: Error: L6218E: Undefined symbol Load$$LR$$LR_m_text$$Limit (referred from BUILD/tests/DISCO_L475VG_IOT01A/ARM/TESTS/mbed_drivers/flashiap/main.o).

davidsaada commented:
@jeromecoutant Yes I am able to reproduce this problem with the ARM compiler (the other toolchains seem to behave properly). Will probably need more time to investigate.

davidsaada commented:
@jeromecoutant Please retest with the ARM compiler. Made a change that depends on another open PR, so even if it works, it'll be some time until it gets merged. You only need to pull TESTS/mbed_drivers/flashiap/main.cpp (other files aren't relevant for your target).

jeromecoutant commented:
Hi
- ARM: OK
- GCC: OK
- IAR: test crashes (it doesn't detect the end of code section ?)

davidsaada commented:
> IAR: test crashes (it doesn't detect the end of code section ?)

OK, I think I understand the problem (has to do with the beginning of the text section). Pushed a fix for it now. Keeps working on the boards I have.
@jeromecoutant Would appreciate retesting. I added a debug print in the beginning of the init test. Will be used if the test doesn't pass and removed if it does.

jeromecoutant reviewed (dismissed):
As discussed by mail, test seems OK with ARM, GCC and IAR now

davidsaada commented:
Thanks @jeromecoutant. Just pushed a commit without the debug print.

0xc0170 commented:
/morph build

cmonr commented:
@davidsaada Please review the build linker issues.

davidsaada commented:
> @davidsaada Please review the build linker issues.

I'm on it (though probably won't get sorted out before next week start).

cmonr commented:
@davidsaada Any progress?

davidsaada commented:
> @davidsaada Any progress?

Currently in touch with Toshiba regarding the link failure in the TMP46M board over IAR. Caused by an obscurity in the IAR linker script for this board, which we're trying to figure. Otherwise everything else is fixed, including the failure in the other board (didn't push it yet).

0xc0170 commented:
Please let us know once updated

cmonr commented:
@davidsaada Any progress since it's been over a month?

davidsaada commented:
Pushed fixes to resolve failures and to handle [#7388](https://github.com/ARMmbed/mbed-os/issues/7388) (same with NVStore test) as well.

0xc0170 commented on drivers/FlashIAP.h:
```diff
@@ -28,6 +28,20 @@
 #include "platform/NonCopyable.h"
+#include <algorithm>
+
+// Export ROM end address
```
This is part of the fixing tests, introducing symbols via FlashIAP to user space, is this intentional? It might be worth do this as separate commits with a reason why are these being added (is this just for testing?)

`FlashIAP: add ROM end macro` would be sufficient? If we do not already provide these anywhere

davidsaada replied:
We don't provide this anywhere else and it's actually quite a good idea to split it. I can think of NVStore (not the test but the feature itself) using this for validation. Will split it into a separate PR.

davidsaada replied:
Done. Depends on [#7705](https://github.com/ARMmbed/mbed-os/pull/7705) now.

davidsaada commented:
Rebased. No longer depending on any other PR.

cmonr commented:
/morph build

cmonr commented:
@davidsaada Please take a look at the build failures.

davidsaada commented:
> @davidsaada Please take a look at the build failures.

Ugh... Apparently this PR is in a race with the addition of new boards. Will open another PR similar to [#7242](https://github.com/ARMmbed/mbed-os/pull/7242) to fix this.

davidsaada commented:
Depending now on [#7729](https://github.com/ARMmbed/mbed-os/pull/7729) to fix ARM linker issues.

davidsaada commented:
Rebased. No longer depending on any other PR.

0xc0170 commented on targets/TARGET_TOSHIBA/TARGET_TMPM46B/device/TOOLCHAIN_IAR/tmpm46bf10fg.icf:
```diff
@@ -25,11 +25,11 @@ define region RAM_region   = mem:[from __ICFEDIT_region_RAM_start__   to __ICFED
 
-define block FLASH_CODE_ROM  {section FLASH_ROM_init object flash_api.o, section .text_init object tmpm46b_fc.o}; 
-define block FLASH_CODE_RAM  {section FLASH_ROM object flash_api.o, section .text object tmpm46b_fc.o};
+define block FLASH_CODE_ROM  {section FLASH_ROM_init object flash_api.o};
```
what is this change about? removing some temporary object that should not be there? Separate commit for this change with reason for removing?

davidsaada replied:
This was modified after a the previous link failure in this board. Consulted with Toshiba engineers, and they said that this object file wasn't required there. Now, I can move this change to another PR, but it would only make sense in this one, as the link error was the only reason for this change.

davidsaada replied:
@0xc0170 Is this explanation satisfactory? If not, I will need to open another PR (with the reasoning behind it being the resolving of the link error in this one).

cmonr replied:
@davidsaada I would recommend backing out this change and moving it to another PR since it's not really related to the PR.

davidsaada replied:
OK Done. Now depending on [#7791](https://github.com/ARMmbed/mbed-os/pull/7791).

davidsaada commented:
(As it's weekend, and want to know if I'm able to split the the PRs)
/morph build.

cmonr commented:
@davidsaada Sneaky sneaky.

The one test timeout failure has been fixed with another PR. As of right now, it fails intermittently.

cmonr commented:
/morph build

cmonr commented:
@ARMmbed/mbed-os-storage Mind taking a quick look at the test changes? Should be fine.

offirko reviewed (approved):
Looks great to me

davidsaada commented:
Test needs to be rerun. Failure is on a timeout after flashing the image. Tried myself with the same board and it works OK.

cmonr commented:
/morph test

cmonr commented:
Welp, at least now the test that caused this PR to fail is not longer in the repo.

/morph test

cmonr commented:
/morph test

davidsaada commented:
@cmonr  Rebased following merge of [#7670](https://github.com/ARMmbed/mbed-os/pull/7670) (as both modify the NVstore tests). Please rerun morph build when done.

cmonr commented:
/morph build

cmonr commented:
/morph mbed2-build

davidsaada commented:
@cmonr Don't know what this failure is about, probably not related to this PR. Regardless, please note that this PR is depending on [#7791](https://github.com/ARMmbed/mbed-os/pull/7791). As both PRs are on the verge of merging (after uVisor test failure gets sorted out), I don't think we should rebase this one and go through the entire morph build process again. Please just take this into consideration if and when both PRs get merged.

cmonr commented:
It's possible that a commit that was merged caused the uVisor test to break in a strange way. We've been keeping an eye on it this weekend since it's blocking a whole slew of PRs from being merged :(

NirSonnenschein commented:
/morph uvisor-test

cmonr commented:
Retagging for 5.10 since FLASHIAP_ROM_END is not in a patch release.
